compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -138,6 +350,90 @@ emitLinkedTuningSpec(ModuleOp module, ArrayRef<NamedSequenceOp> specsToLink) {
   return newSpec;
 }
 
+static FailureOr<NamedSequenceOp> emitLinkedDefaultTuningSpec(ModuleOp module) {
+  OpBuilder builder(module.getContext());
+
+  // Step 1: Collect NamedSequenceOps from inner modules.
+  TuningSpecsToMerge tuningSpecs = collectTuningSpecsToMerge(module);
+  SmallVector<NamedSequenceOp> &namedSequenceOpsToMove =
+      tuningSpecs.namedSequenceOpsToMove;
+  llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> &namedSequenceToUser =
+      tuningSpecs.namedSequenceToUser;
+
+  // Step 2-a: Make sure the name sequence names are unique, and then move
+  // collected NamedSequenceOps to the top-level module.
+  resolveAndMoveNamedSequenceOps(namedSequenceOpsToMove, module, builder,
+                                 namedSequenceToUser);
+
+  // Step 2-b: Create a new NamedSequenceOp `__kernel_config` in the top-level
+  // module.
+
+  // Collect all `ForeachMatchOp`s from `__kernel_config` NamedSequenceOps
+  // in inner modules to merge into the new `__kernel_config` in the top module.
+  SmallVector<ForeachMatchOp> foreachMatchOps;
+  for (auto innerModule : module.getBody()->getOps<ModuleOp>()) {
+    // All other NamedSequenceOps have already been moved to the top-level
+    // module in Step 2-a. As a result, each inner module now contains only one
+    // `__kernel_config` NamedSequenceOp that is annotated with
+    // `iree_codegen.tuning_spec_entrypoint`.
+    auto namedSequenceOp =
+        getSingleElement(innerModule.getOps<NamedSequenceOp>());
+    ForeachMatchOp foreachMatch =
+        getForeachMatchOpFromKernelConfig(namedSequenceOp);
+    foreachMatchOps.push_back(foreachMatch);
+  }
+
+  builder.setInsertionPointToEnd(module.getBody());
+  Location loc = module.getLoc();
+  NamedSequenceOp newNamedSequence =
+      createKernelConfigOp(builder, loc, kKernelConfigSpecName);
+
+  bool hasConsumedArg = llvm::any_of(foreachMatchOps, hasConsumedArgument);
+  StringRef attrName =
+      hasConsumedArg ? kArgConsumedAttrName : kArgReadOnlyAttrName;
+  newNamedSequence.setArgAttr(0, attrName, builder.getUnitAttr());
+  newNamedSequence->setAttr(kTuningSpecEntrypointAttrName,
+                            builder.getUnitAttr());
+  // Indicate that the output module is a default tuning spec after merging.
+  module->setAttr(kTuningSpecDefaultEntrypointAttrName, builder.getUnitAttr());
+
+  // Step 2-c: Create a new block inside the NamedSequenceOp and merge the
+  // ForeachMatchOp from each inner module into one ForachMatchOp.
+  Type anyOpType = builder.getType<transform::AnyOpType>();
+  SmallVector<Type, 4> resultTypes = {anyOpType};
+  SmallVector<Attribute> mergedMatchers;
+  SmallVector<Attribute> mergedActions;
+
+  for (auto foreachMatchOp : foreachMatchOps) {
```

Comment:
Use the actual type here, since it's not obvious based on the context: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable